### PR TITLE
Fix/galley sns length

### PR DESCRIPTION
**Pull request by Evoludo**
Fixes uncaught errors from SNS when payload is too large. Token creation in tests is (hopefully) a bit clearer now too.

I've also included a few fixes in the Makefiles for building the integration tests, and updated the brig schema so it can be brought up in Cassandra 3.

### CODE REVIEW

File: services/brig/Makefile [hunk 1]
```diff
@@ -28,6 +28,7 @@ init:
 .PHONY: install
 install: init
 	stack install --pedantic --test --local-bin-path=dist
+	cp "$(shell stack path --dist-dir)/build/$(NAME)-integration/$(NAME)-integration" dist/
```
Comment (elverkilde) [inline services/brig/Makefile]: bad rebase?

File: services/brig/schema/src/Main.hs [hunk 2]
```diff
@@ -23,7 +23,6 @@ import qualified V21
 import qualified V22
 import qualified V23
 import qualified V24
-import qualified V25
```
Comment (elverkilde) [inline services/brig/schema/src/Main.hs]: is it safe to remove this?

File: services/brig/schema/src/V18.hs [hunk 3]
```diff
@@ -12,7 +12,7 @@ migration = Migration 18 "Add prekeys" $ do
     void $ schema' [r|
         create columnfamily if not exists clients
             ( user   uuid
-            , client ascii
+            , client text
```
Comment (elverkilde) [inline services/brig/schema/src/V18.hs]: should we put these changes in a separate PR maybe?

File: services/gundeck/src/Gundeck/Aws.hs [hunk 4]
```diff
@@ -91,6 +92,8 @@ import qualified Network.AWS.SQS         as SQS
 import qualified Network.TLS             as TLS
 import qualified System.Logger           as Logger
 
+import Debug.Trace
```
Comment (elverkilde) [inline services/gundeck/src/Gundeck/Aws.hs]: oops

File: services/gundeck/src/Gundeck/Aws.hs [hunk 5]
```diff
@@ -306,6 +316,12 @@ createEndpoint u tr env app token = do
         _ <- string "Invalid parameter: Token"
         return ()
 
+    isLengthError Nothing = False
+    isLengthError (Just s) = isRight . flip parseOnly (toText s) $ do
+        _ <-  (string "Invalid parameter: Token Reason: must be at most 8192 bytes long in UTF-8 encoding")
```
Comment (elverkilde) [inline services/gundeck/src/Gundeck/Aws.hs]: maybe put the prefix string in a let binding to keep the line width smaller

File: services/gundeck/src/Gundeck/Push.hs [hunk 6]
```diff
@@ -235,6 +235,9 @@ addToken (uid ::: cid ::: req ::: _) = do
                 Log.info $ "token" .= tokenText tok
                         ~~ msg (val "Invalid push token.")
                 return (Left invalidToken)
+            Left (Aws.TokenTooLong l) -> do
+                Log.info $ msg ("Push token is too long: token length = " ++ show l)
```
Comment (elverkilde) [inline services/gundeck/src/Gundeck/Push.hs]: remove debug logging

File: services/brig/schema/src/Main.hs (hunk 2 above)
Comment (Evoludo) [reply services/brig/schema/src/Main.hs]: Yes, this was the version that migrated the the schema form ascii to text (this isn't allowed in Cassandra 3, which we now use). Since the schema now starts off in text. this is no longer required. There's no other way to bring up this schema in Cassandra 3.

File: services/brig/Makefile (hunk 1 above)
Comment (Evoludo) [reply services/brig/Makefile]: Possibly, does this live in another branch?

File: services/gundeck/src/Gundeck/Aws.hs (hunk 4 above)
Comment (Evoludo) [reply services/gundeck/src/Gundeck/Aws.hs]: Whoops, thought I removed that!

File: services/gundeck/src/Gundeck/Push.hs (hunk 6 above)
Comment (Evoludo) [reply services/gundeck/src/Gundeck/Push.hs]: Are you sure? The other errors caught here have it.

File: services/brig/schema/src/V18.hs (hunk 3 above)
Comment (Evoludo) [reply services/brig/schema/src/V18.hs]: On thinking about it again, that's probably best, yes.

File: services/gundeck/test/integration/API.hs [hunk 7]
```diff
@@ -684,6 +685,27 @@ testReplaceSharedPushToken g b = do
...
+testLongPushToken g b = do
+    uid <- randomUser b
+    clt <- randomClient g uid
+    
+    -- normal size APNS token should succeed
+    tkn1 <- randomApnsToken clt defTS
+    registerPushTokenRequest uid tkn1 g !!! const 201 === statusCode
+
+    -- APNS token over 400 bytes should fail (actual token sizes are twice the tSize)
+    tkn2 <- randomApnsToken clt defTS{tSize=256}
```
Comment (jschaul) [inline services/gundeck/test/integration/API.hs]: `defTS{tSize=256}` seems an odd thing to do, since `defTS` is `TokenSpec 32 appName`, i.e. already specifies a length of 32, which you then override again. How about simply using `(TokenSpec 256 appName)` instead?

File: services/gundeck/src/Gundeck/Aws.hs [hunk 8]
```diff
@@ -293,7 +299,9 @@ createEndpoint u tr env app token = do
   where
     readArn r = either (throwM . InvalidArn r) return (fromText r)
 
-    -- Thank you Amazon for not having granular error codes!
+    tokenLength = toInteger . Text.length . Push.tokenText
+
+    -- Thank you Amazon for not having granul\ar error codes!
```
Comment (jschaul) [inline services/gundeck/src/Gundeck/Aws.hs]: `granular` -> `granul\ar` ?

File: services/gundeck/test/integration/API.hs [hunk 9]
```diff
@@ -810,16 +836,30 @@ buildPush sdr rcps pload =
...
+randomToken :: MonadIO m => ClientId -> Int -> Transport -> AppName -> m PushToken
+randomToken c size trans name = liftIO $ do
     tok <- Token . T.decodeUtf8 <$> case trans of
-        GCM -> toByteString' <$> randomId
-        _   -> B16.encode    <$> randomBytes 32
+        GCM -> multipleUuids size randomId
+        _   -> B16.encode    <$> randomBytes size
     return (pushToken trans name tok c)
 
+multipleUuids :: Control.Monad.IO.Class.MonadIO m => Int -> m (Id a) -> m ByteString
```
[the remark this answers is not in the document]
Comment (Evoludo) [reply services/gundeck/test/integration/API.hs]: Done!

File: services/gundeck/test/integration/API.hs (hunk 7 above)
Comment (Evoludo) [reply services/gundeck/test/integration/API.hs]: I did it this way so that a default TokenSpec could be specified, and have arbitrary parameters overridden, instead of statically specifying the same values again for the same use cases.
